api/envoy/extensions/clusters/dns/v3/dns_cluster.proto

```diff
@@ -81,7 +84,6 @@ message DnsCluster {
   // :ref:`AUTO<envoy_v3_api_enum_value_extensions.clusters.common.dns.v3.DnsLookupFamily.AUTO>`.
   common.dns.v3.DnsLookupFamily dns_lookup_family = 8;
 
-  // Required. Whether to do logical or strict dns resolution.
-  // This field is only considered when the ``name`` field of the ``TypedConfig`` is ``envoy.cluster.dns``.
-  DnsDiscoveryType dns_discovery_type = 9;
+  // If true, perform logical DNS resolution. Otherwise, perform strict DNS resolution.
+  bool logical = 9;
```

Comment:
I agree that "logical" is confusing, but I also don't see how "all_addresses_in_single_endpoint" captures the semantics of logical vs strict dns clusters. I could be wrong, but my understanding is as follows:

> [logical dns] is optimal for large scale web services that must be accessed via DNS. Such services typically use round robin DNS to return many different IP addresses. Typically a different result is returned for each query. If strict DNS were used in this scenario, Envoy would assume that the cluster’s members were changing during every resolution interval which would lead to draining connection pools, connection cycling, etc. Instead, with logical DNS, connections stay alive until they get cycled.

It seems to me that the crucial difference between is how existing hosts are treated after a dns query, not whether all addresses are in a single endpoint (the [docs](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/service_discovery#logical-dns) don't even mention the latter). Some people may even want strict dns resolution when all addresses are in a single endpoint.

Reply:
I'm not familiar with the Envoy implementation, but from having implemented LOGICAL_DNS support in gRPC, I think the key difference is actually the fact that all addresses are considered part of a single endpoint.

In a LOGICAL_DNS cluster, Envoy creates a single Host (which is the object it uses internally to represent an endpoint) with all addresses.  Since there is only one Host, all requests will always be sent to that `Host`, regardless of which LB policy is used.  Every time the Host is asked for a new connection, it uses whichever address is currently at the front of the list.

In a STRICT_DNS cluster, we create a separate Host for each address.  The LB policy is used to pick which Host each request should be sent to, and the chosen Host manages connections to that address.

@mattklein123 can correct me if I'm wrong here, but the above is my understanding.

Reply:
@markdroth I think I see what you mean. Logical DNS clusters have a single `Host` with one address. This address may change based on results of DNS queries.

My question for @mattklein123 (or anyone familiar with the matter) is why the requirement for there to be a single host? Is there a fundamental reason why we can't have a config like

```yaml
      name: my_cluster
      type: LOGICAL_DNS
      lb_policy: ROUND_ROBIN 
      hosts:
        - socket_address:
            address: eastus.service.com
            port_value: 80
        - socket_address:
            address: westus.service.com
            port_value: 80       
```

This would create two `Host`s under the hood, one for `eastus.service.com` and one for `westus.service.com`, and envoy would round robin between them. When any `Host` is asked for an address, it would return the one "at the front of [its] list".

Reply:
In principle, I don't see any reason why we couldn't add a mechanism to do something like that, but it's not something we have today, and we probably wouldn't want to actually do that without a concrete requirement.

But I also think that that feature isn't really relevant to the question of how we indicate the difference between LOGICAL_DNS and STRICT_DNS clusters.  If we were going to introduce the ability for a single DNS cluster to look up multiple names, we could do it for both types, but there would still be the same difference in behavior between LOGICAL_DNS and STRICT_DNS: I would expect STRICT_DNS to still create a separate `Host` for each address, and we could define the LOGICAL_DNS behavior to be either (a) have one `Host` for each name, each of which has all of the addresses for that name, or (b) have a single `Host` containing all addresses for both names -- which behavior we chose would depend on the concrete use-case we'd be trying to address.

Reply:
@markdroth Makes sense

Reply:
The point of LOGICAL_DNS is to work with DNS that points to very large web services (think AWS S3 or something like that). You don't care who you send to as you assume they are doing load balancing. This is very different from something like K8s DNS based load load balancing. Both have there purpose.

And sorry I haven't really been tracking this PR. Can you please ping me optimally on Slack when this is ready for final review?

Reply:
Sure that seems fine if there is no other difference.

Reply:
From the perspective of the xDS data model and familiarity with the gRPC implementation, I think that's the only difference.  However, I'm not familiar with Envoy's implementation, which is why I was asking for confirmation.

If you don't know, is there someone else we can ask who is more familiar with the Envoy implementation?  I just want to make sure we're not introducing a misleading API here.

Thanks!

Reply:
I would have to page it all back in and look. I can try to find some time for that.

Reply:
Unless anyone can think of any reason why `bool all_addresses_in_single_endpoint` doesn't work, I'd like to see this changed.  I really don't think `bool logical` makes sense as an API here.

Reply:
@markdroth I think we've resolved the last outstanding item. With your and @mattklein123 's approval, I'm ready to merge whenever.